gdal/frmts/gtiff/cogdriver.cpp

```diff
@@ -1011,6 +1011,36 @@ GDALDataset* GDALCOGCreator::Create(const char * pszFilename,
          }
     }
 
+    const char* osOvrCompress = CSLFetchNameValueDef(papszOptions, "OVERVIEW_COMPRESS", "AUTO");
+    if ( !EQUAL(osOvrCompress, "AUTO") ) 
+    {
+        CPLSetConfigOption("COMPRESS_OVERVIEW", osOvrCompress);
```

Comment:
Please use CPLConfigOptionSetter (cpl_conv.h) for a RAAI solution (the setter object must be put in a higher scope) that uses thread-local config options

Reply:
You'll probably have to use a std::unique_ptr<CPLConfigOptionSetter>

Reply:
Since they are just pass through values, I could just pass them directly through,

for example, I could replace
```
    const char* osOvrQuality = CSLFetchNameValueDef(papszOptions, "OVERVIEW_QUALITY", "AUTO");
    if ( !EQUAL(osOvrQuality, "AUTO") ) 
    {
        CPLSetConfigOption("JPEG_QUALITY_OVERVIEW", osOvrCompress);
        CPLSetConfigOption("WEBP_LEVEL_OVERVIEW", osOvrCompress);
    }
```
with 
```
    CPLConfigOptionSetter ovrQualityJpegSetter("JPEG_QUALITY_OVERVIEW", CSLFetchNameValueDef(papszOptions, "OVERVIEW_QUALITY", "75"), true);
    CPLConfigOptionSetter ovrQualityWebpSetter("WEBP_LEVEL_OVERVIEW", CSLFetchNameValueDef(papszOptions, "OVERVIEW_QUALITY", "75"), true);
```

And then remove the option of `"AUTO"`

otherwise i need to read some more about deconstructors and blocks scoping in c++

Reply:
> for a RAAI solution 

also i am not sure what "RAAI" means in this context, Its been many years since Ive done any serious c/c++ programming, the internet thinks its "Resource Acquisition Is Initialization"  https://en.cppreference.com/w/cpp/language/raii which makes sense when we are talking about `CPLConfigOptionSetter`

Reply:
Just to be sure: is photometric still set automatically into YCBCR for JPEG compressed 3 band overviews? Even if the full resolution has some other compression?

Reply:
> RAAI

Yes I meant RAII

>  is photometric still set automatically into YCBCR for JPEG compressed 3 band overviews? Even if the full resolution has some other compression?

good point. I don't think so currently. GTIFFWriteDirectory() is called with the m_nPhotometric value from the full resolution image. CreateOverviewsFromSrcOverviews() should also read PHOTOMETRIC_OVERVIEW, RegisterNewOverviewDataset allow to override m_nPhometric in the overview dataset and the COG driver to automatically set PHOTOMETRIC_OVERVIEW=YCBCR for OVERVIEW_COMPRESS=JPEG

Reply:
I am not too familiar with the inner workings of JPEG and `YCBCR`, but should it be set automatically? are there any cases when the other types of `PHOTOMETRIC` should be included?

Looking at how the `PHOTOMETRIC` creation option is handled there is also a bunch of extra logic for other types eg `CMKYK` or `RGB` would they also need to be handled too? 

Is defaulting to `YCBCR` overviews for all COGs that have `OVERVIEW_COMPRESS=JPEG`  what you want to do?

Reply:
> Is defaulting to `YCBCR` overviews for all COGs that have `OVERVIEW_COMPRESS=JPEG` what you want to do?

yes that's what we do for the full resolution imagery in the COG driver already, so doing it for overviews is consistent with that.